compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1987,13 +1987,49 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         // the fields with the base_expr. This could cause us to hit errors later
         // when certain fields are assumed to exist that in fact do not.
         if error_happened {
-            if let Some(base_expr) = base_expr {
+            if let hir::Rest::Base(base_expr) = base_expr {
                 self.check_expr(base_expr);
             }
             return;
         }
 
-        if let Some(base_expr) = base_expr {
+        if let hir::Rest::DefaultFields(span) = *base_expr {
+            for f in &variant.fields {
+                let ident = self.tcx.adjust_ident(f.ident(self.tcx), variant.def_id);
+                if let Some(_) = remaining_fields.remove(&ident)
+                    && f.value.is_none()
+                    && self.tcx.features().default_field_values()
+                {
+                    let guar = self
+                        .dcx()
+                        .struct_span_err(span, format!("missing mandatory field `{ident}`"))
+                        .emit();
+                    self.set_tainted_by_errors(guar);
+                    return;
+                }
```

Comment:
This should collect *all* the missing mandatory fields and emit a single error mentioning all of them (and their types) before returning.